### PR TITLE
Refactor train_batch

### PR DESCRIPTION
**Pull request by sakshigupta93**
# What does this PR do?
Refactor out the train_batch() code to have everything between the batch start and end events. Exposes this as an external function so that it can be called independently.

# What issue(s) does this change relate to?
https://docs.google.com/document/d/1XbPux2ISP-zOyJhflvmAf_T3BeDgpIOyCbWOyNeQ0Sw/edit?pli=1#heading=h.4cz970y1mk93

Testing: Started a corgi run, will verify that it succeeds. composer-regression-tests-runner-k745y0
All E2E tests are passing.

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -2209,7 +2146,7 @@ def _run_evaluators(self, event: Event):
 
         self.engine.run_event(Event.EVAL_AFTER_ALL)
 
-    def _train_batch(self, use_grad_scaling: bool) -> Dict[str, torch.Tensor]:
+    def train_batch(self, batch: Any, use_grad_scaling: bool, last_wct: Optional[Any]) -> Any:
```
Comment (b-chu) [inline composer/trainer/trainer.py:2149]: last_wct should be typed

File: composer/trainer/trainer.py [hunk 2]
```diff
@@ -2209,7 +2146,7 @@ def _run_evaluators(self, event: Event):
...
         """Compute loss by training on a full batch of data.
```
Comment (b-chu) [inline composer/trainer/trainer.py:2150]: Improve this docstring by adding args, returns, and the new functionality

**Review by b-chu — CHANGES_REQUESTED** [review]
Thanks for the PR!

**Review by b-chu — APPROVED** [review]
Thanks for the PR Sakshi!

**Comment by eracah** [comment]
Also I think there are some setup items in fit that we need to add to train batch that get called the first time the user calls fit

**Pull request closed without merging**